File: DataStructure/Trees/Trie.ts
class TrieNode {
  #data: string;
  #children: TrieNode[];
  isEnd: boolean;

  // 假设只保存26个英文字母
  constructor(data: string = "") {
    this.#children = new Array(26);
    this.isEnd = false;
    this.#data = data;
  }

  get children() {
    return this.#children;
  }

  get data() {
    return this.#data;
  }
}

export class Trie {
  #root: TrieNode;

  constructor() {
    this.#root = new TrieNode();
  }

  public insert(word: string) {
    word = word.toLowerCase();
    let node = this.#root;
    for (let i = 0; i < word.length; i++) {
      const index = word.charCodeAt(i) - 97;
      if (node.children[index] === undefined) {
        node.children[index] = new TrieNode(word.charAt(i));
      }
      node = node.children[index];
    }
    node.isEnd = true;
  }

  public delete(word: string) {
    let node = this.#root;
    for (let i = 0; i < word.length; i++) {
      const index = word.charCodeAt(i) - 97;
      if (node.children[index] === undefined) return;
      node = node.children[index];
      console.log(node.data);
    }
    if (node) {
      node.isEnd = false;
    }
  }

  public contains(word: string) {
    let node = this.#root;
    for (let i = 0; i < word.length; i++) {
      const index = word.charCodeAt(i) - 97;
      if (node.children[index] === undefined) return false;
      node = node.children[index];
    }
    return node.isEnd;
  }

  public startsWith(prefix: string) {
    let node = this.#root;
    for (let i = 0; i < prefix.length; i++) {
      const index = prefix.charCodeAt(i) - 97;
      if (node.children[index] === undefined) return false;
      node = node.children[index];
    }
    return node !== undefined;
  }
}

File: DataStructure/Trees/Trie.go
package main

import "fmt"

// AlphabetSize is the number of possible characters in the trie
const AlphabetSize = 26

// Node
type Node struct {
	children [26]*Node
	isEnd    bool
}

// Trie
type trie struct {
	root *Node
}

func New() *trie {
	return &trie{root: &Node{}}
}

// Insert
func (t *trie) Insert(word string) {
	wordLength := len(word)
	currentNode := t.root
	for i := 0; i < wordLength; i++ {
		charIndex := word[i] - 'a'
		if currentNode.children[charIndex] == nil {
			currentNode.children[charIndex] = &Node{}
		}
		currentNode = currentNode.children[charIndex]
	}
	currentNode.isEnd = true
}

// Search
func (t *trie) Search(word string) bool {
	wordLength := len(word)
	currentNode := t.root
	for i := 0; i < wordLength; i++ {
		charIndex := word[i] - 'a'
		if currentNode.children[charIndex] == nil {
			return false
		}
		currentNode = currentNode.children[charIndex]
	}
	return currentNode.isEnd == true
}

func main() {
	trie := New()
	trie.Insert("abcd")
	fmt.Println(trie.Search("ab"))
	fmt.Println(trie.Search("abc"))
	fmt.Println(trie.Search("abcd"))
}
